### ISSUE TITLE
Starting in v3.8.0, strategic merge patches append to arrays (earlier versions prepend)

### ISSUE THREAD

**Issue by ephesused** [kubernetes-sigs/kustomize#2907]
Prior to v3.8.0, strategic merge patches placed new elements at the front of arrays. Starting with v3.8.0 - including current master - strategic merge patches place new elements at the end of arrays. This behavior change means that it's possible to take content that works as expected under v3.7.0 and kustomize fail under 3.8.x. It also is possible to have content that works as expected under v3.7.0, has a successful kustomize build, but yields different results under 3.8x.

The meaningful difference arises when such a strategic patch merge is followed by a json patch that alters an element according to its position in the array. For v3.7.0, one can rely on a strategic merge patch addition at the 0th element. For v3.8.0, the newly added element now appears elsewhere. The result is that the following json patch will apply to a different object than it did in v3.7.0, which can mean the patch application will not arrive where it was intended - in some cases, that difference can lead to a failure if the requested path no longer exists.

The result is that moving from v3.7.0 to a subsequent release requires detailed evaluation to ensure proper behavior when strategic merge patches are applied.

The [v3.8.0 release announcement](https://github.com/kubernetes-sigs/kustomize/releases/tag/kustomize%2Fv3.8.0) does state that "The output may order fields within objects differently...", but the reordering here is for arrays. As such, I'm submitting this issue as it "looks not just different, but incorrect." I can't say definitively that the new behavior is incorrect, but it certainly is a meaningful change from prior versions.

Apologies for the last of a test case PR.

### Test setup

kustomization.yaml

```yaml
resources:
  - resources.yaml

patchesStrategicMerge:
  - merge.yaml

#transformers:
#  - transformers.yaml
```

resources.yaml

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: deployment
spec:
  template:
    spec:
      containers:
      - name: existing-container
        image: existing-image
        env: []
```

merge.yaml

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: deployment
spec:
  template:
    spec:
      containers:
      - name: new-container
        image: new-image
```

transformers.yaml

```yaml
apiVersion: builtin
kind: PatchTransformer
metadata:
  name: add-env
patch: |-
  - op: add
    path: /spec/template/spec/containers/1/env/-
    value:
      name: NEW-KEY
      value: NEW-VALUE
target:
  kind: Deployment
```

### Expected behavior (from v3.7.0)

new-container is added at the start of the containers array.

```
[root@8f6da0e966cf project_root]# build/kustomize-execs/release-v3.7.0/kustomize version
{Version:kustomize/v3.7.0 GitCommit:42d1f7b792a80af34828ec4c89af99e1043351a7 BuildDate:2020-07-04T19:15:46Z GoOs:linux GoArch:amd64}
[root@8f6da0e966cf project_root]# build/kustomize-execs/release-v3.7.0/kustomize build testdata/kust-issue-tbd
apiVersion: apps/v1
kind: Deployment
metadata:
  name: deployment
spec:
  template:
    spec:
      containers:
      - image: new-image
        name: new-container
      - env: []
        image: existing-image
        name: existing-container
[root@8f6da0e966cf project_root]#
```

### Actual behavior (v3.8.0 and above, including master)

new-container is added at the end of the containers array.

```
[root@8f6da0e966cf project_root]# build/kustomize-execs/release-v3.8.0/kustomize version
{Version:kustomize/v3.8.0 GitCommit:6a50372dd5686df22750b0c729adaf369fbf193c BuildDate:2020-07-05T14:08:42Z GoOs:linux GoArch:amd64}
[root@8f6da0e966cf project_root]# build/kustomize-execs/release-v3.8.0/kustomize build testdata/kust-issue-tbd
apiVersion: apps/v1
kind: Deployment
metadata:
  name: deployment
spec:
  template:
    spec:
      containers:
      - image: existing-image
        name: existing-container
      - image: new-image
        name: new-container
[root@8f6da0e966cf project_root]#
```

### Failure condition and unexpected application

The append vs. prepend behavior can lead to failure or unexpected patch application results, since the object at at given index is different in v3.7.0 versus v3.8.0 and above..

For the failure case, uncomment the transformers from kustomization.yaml. In v3.7.0, the container at index 1 has an env array; in v3.8.0, the container at index 1 has no env array.

```
[root@8f6da0e966cf project_root]# build/kustomize-execs/release-v3.7.0/kustomize build testdata/kust-issue-tbd
apiVersion: apps/v1
kind: Deployment
metadata:
  name: deployment
spec:
  template:
    spec:
      containers:
      - image: new-image
        name: new-container
      - env:
        - name: NEW-KEY
          value: NEW-VALUE
        image: existing-image
        name: existing-container
[root@8f6da0e966cf project_root]# build/kustomize-execs/release-v3.8.0/kustomize build testdata/kust-issue-tbd
Error: add operation does not apply: doc is missing path: "/spec/template/spec/containers/1/env/-": missing value
[root@8f6da0e966cf project_root]#
```

In the same vein, situations exist where kustomize succeeds but the patch application is on a different object than intended. Following the failure condition edit, also edit merge.yaml and add an empty env array to new-container.

Note how v3.7.0 applies NEW-KEY to existing-container, whereas v3.8.0 applies NEW-KEY to new-container.

```
[root@8f6da0e966cf project_root]# build/kustomize-execs/release-v3.7.0/kustomize build testdata/kust-issue-tbd
apiVersion: apps/v1
kind: Deployment
metadata:
  name: deployment
spec:
  template:
    spec:
      containers:
      - env: []
        image: new-image
        name: new-container
      - env:
        - name: NEW-KEY
          value: NEW-VALUE
        image: existing-image
        name: existing-container
[root@8f6da0e966cf project_root]# build/kustomize-execs/release-v3.8.0/kustomize build testdata/kust-issue-tbd
apiVersion: apps/v1
kind: Deployment
metadata:
  name: deployment
spec:
  template:
    spec:
      containers:
      - image: existing-image
        name: existing-container
      - env:
        - name: NEW-KEY
          value: NEW-VALUE
        image: new-image
        name: new-container
[root@8f6da0e966cf project_root]#
```

**Comment by Shell32-Natsu** [comment]
Well, this is another behavior change due to migrating to kyaml. In my opinion, append to the array is more intuitive than prepend.

**Comment by ephesused** [comment]
My concern is more that we cannot use the same set of inputs for v3.7.0 (and earlier) that we use for v3.8.0 (and later). The breaking change means our inputs are not compatible across kustomize versions.

**Comment by seh** [comment]
I happened to run into this last Friday, not working with _kustomize_, but just with _kubectl patch_. I couldn't find any documentation that specifies what Strategic Merge Patch does when merging into a list. My case involves adding a new init container, where order is crucial. I found by way of experiments that _kubectl patch_ appends (using version 1.19.2). I needed to add the new init container to the beginning, so I had to fall back to JSON Patch instead.
